Support more ways to load kernel modules early

Foxboron opened the pull request:
This path implements two new ways of loading kernel modules early. Sometimes building monolithic kernels are not really feasible and people use their kernel and kernel modules included int he initrd image. This brings problems like how to supply the desired driver into `/lib/modules/*.ko` as you would need to have an explicit `-file` parameter for each module.

This introduces two new ways to get around this:
* Through configuration files in `/etc/modules-load.d`
* Through `modules_load` kernel parameter

These options are popularized through the systemd project and documented here;

https://www.freedesktop.org/software/systemd/man/kernel-command-line.html#modules_load=
https://www.freedesktop.org/software/systemd/man/modules-load.d.html

I'm also wondering if we'd be interested in supporting a new u-root argument `-module` which would create a `/etc/modules-load.d/default.conf` file in the initrd, but this can come later.

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -238,7 +239,10 @@ var excludedMods = map[string]bool{
 }
 
-// InstallAllModules installs kernel modules (.ko files) from /lib/modules.
+// InstallAllModules installs kernel modules form the following locations in order;
```
"form" is a typo

abrender replied:
"in order:" instead of "in order;"

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+	r := bufio.NewReader(f)
+	modules := []string{}
+	for {
+		if i, err := r.ReadString('\n'); err == nil {
```
Does it make sense to use strings.TrimSpace to remove leading and trailing whitespace?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+	modules := []string{}
+	for {
+		if i, err := r.ReadString('\n'); err == nil {
+			if strings.HasPrefix(i, "#") {
```
do you also want to check for empty lines?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+				modules = append(modules, strings.TrimSuffix(i, "\n"))
+			}
+		} else {
+			break
```
does it make sense to log an error if err != io.EOF?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+	for _, filename := range files {
+		f, err := os.Open(filename)
+		if err != nil {
+			log.Printf("installModulesFromModulesLoad: can't open %q: %v", filename, err)
```
The log line prints "installModulesFromModulesLoad" but the function name is "InstallModulesFromModulesLoad" (capital "I" at the start)

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+		fModules := readModules(f)
+		for _, moduleName := range fModules {
+			if _, ok := exclude[moduleName]; ok {
+				log.Printf("Skipping module %s", moduleName)
```
%q instead of %s might help here... in an edge case where the module name is empty it might not be obvious to a user reading the log line that the module name is empty

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+			}
+			flags := cmdline.FlagsForModule(moduleName)
+			err = kmodule.Probe(moduleName, flags)
+			f.Close()
```
should f.Close() be "defer f.Close()" on line 329, right after check of `err` after os.Open?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+			flags := cmdline.FlagsForModule(moduleName)
+			err = kmodule.Probe(moduleName, flags)
+			f.Close()
+			if err != nil {
```
I think you can inline this:
if err := kmodule.Probe(...); err != nil {
 // error here
}

also: the function name in the log message on line 339 is missing the initial uppercase "i"

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+}
+
+// InstallModules installs kernel modules from .conf files in /etc/modules-load.d/
+func InstallModulesFromCmdline(pattern string, exclude map[string]bool) error {
```
"pattern" isn't used in this function... is this argument needed?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+
+	for _, moduleName := range strings.Split(modules, ",") {
+		if _, ok := exclude[moduleName]; ok {
+			log.Printf("Skipping module %s", moduleName)
```
%q instead of %s here please?

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+		return nil
+	}
+
+	for _, moduleName := range strings.Split(modules, ",") {
```
do you have to worry about whitespace? What if someone passes:
--modules_load="module_a, module_b , module_c" ?

Foxboron replied:
I would just call it invalid and move on. But we can strip both sides of the string?

abrender replied:
I think that calling strings.TrimSpace(moduleName) is so easy that we should just do it and help ensure a good experience for users.

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -284,3 +295,71 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+		}
+		flags := cmdline.FlagsForModule(moduleName)
+		err := kmodule.Probe(moduleName, flags)
+		if err != nil {
```
I think you can inline the error checking: `if err := kmodule.Probe(...); err != nil { ...`

abrender commented:
Thank you for this contribution!  Can you please also add test coverage for the code? Thanks!

rminnich commented:
I have always wondered if your module loading setup could follow the standard path:
/lib/modules/kernel-verison/modules.dep for information, and that path for modules.
I don't see the point of /etc/modules.d, as part of the goal of u-root was to avoid bringing in dependencies on /etc, and it is actually redundant with what is in /lib/modules/kernel-version. 

wdyt, am I missing something?

Foxboron commented:
The question currently is; if you have `/lib/modules/$kernel-$ver` and you need to load modules at loadtime. What options do you have?

Currently I can just see ``/lib/modules/*.ko` but no distro package does this, so you'd have to place them there manually. It's more convenient to just include a module directory and have the module resolution as implemented in `kmod` do it's job, right? Then you would only really need some way to tell u-root "load these modules up at statup" and `kmod` does the rest.

We can move away from `/etc/modules-load.d` and use `/lib/modules-load.d` instead.

rminnich commented:
Quick question: I have this:
ls /lib/modules
5.15.0-41-generic  5.15.0-43-generic  5.15.0-46-generic

do you not have that? Maybe I'm about to learn something :-)

rminnich commented:
I may be misunderstanding your comment but anyway:
on boot, 
mount proc
read /proc/version to get a version (needs a small bit of processing)
then look in /lib/modules/version

could that work?

Foxboron commented:
>do you not have that? Maybe I'm about to learn something :-)

I do, but that doesn't load any modules.

>read /proc/version to get a version (needs a small bit of processing) then look in /lib/modules/version
> could that work?

`kmod` already implements this and we use that to find the modules in this implementation. The question is how you pass modulenames during boot.

Foxboron commented:
Moved the module directory from `/etc/modules-load.d` to `/lib/modules-load.d` :)

rminnich commented on pkg/libinit/root_linux.go:
```diff
@@ -263,24 +275,97 @@ func InstallModules(pattern string, exclude map[string]bool) error {
 
+// InstallModules installs kernel modules from .conf files in /etc/modules-load.d/
+func InstallModulesFromCmdline(exclude map[string]bool) error {
+	modules, present := cmdline.Flag("modules_load")
```
Right here is where you can use the new cmdline functions to create fake command lines and parse them.

rminnich replied:
OK, what I'd suggest is that these functions change in a few ways. We might want a struct containing needed information, and one item in that struct could be a *CmdLine created from the cmdline package. That CmdLine can then be mocked up for testing or derived from the real cmdline. This should let us right much better tests.

Foxboron replied:
I'll take a few stabs at this tomorrow. I'm not quite happy with this code as-is so I won't mind do a bit of reworking :)

rminnich replied:
how goes? I'd like to get this in.

Foxboron commented:
Okay, so I have implemented test coverage for the new functionality where we load modules by reading `/proc/cmdline` and files from `/lib/modules-load.d`. This now uses a `InitialModulesLoader` struct which allows us to mock up the kmod probing as well and the `Cmdline` struct. 

I renamed the original `/lib/modules/*.ko` from `InstallModules` to `InstallModulesFromDir` and didn't integate it with `kmodule.Probe` as it is never really utilizing the probing infrastructure and do files instead.

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -263,24 +322,84 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+			}
+		} else if err != io.EOF {
+			log.Println("error on reading:", err)
+		} else {
```
(https://pkg.go.dev/strings#TrimSpace) hmm so we need this because `ReadString returns err != nil if and only if the returned data does not end in delim` and the docs for TrimSpace state  `For simple uses, a Scanner may be more convenient.`

Perhaps it'll be clearer to restructure this code block as:

```
scanner := bufio.NewScanner(f)
for scanner.Scan() { // By default this uses newline delimiters
  // Do something with scanner.Text()
}
if err := scanner.Err(); err != nil {
  // Do something with err
}
```

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -263,24 +322,84 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+		return nil, err
+	}
+	for _, filename := range files {
+		fmt.Println(files)
```
should this `fmt.Println` line be removed? Maybe it was left over from debugging?

abrender commented on pkg/libinit/root_linux_test.go:
```diff
@@ -0,0 +1,121 @@
+
+	expectedModules := []string{"test", "something-test"}
+	InstallModules(loader, expectedModules)
+	if !reflect.DeepEqual(expectedModules, loadedModules) {
```
(here and below) if you use `cmp.Diff` (https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) you get a nice diff, instead of 2 long lists that you have to manually diff

```
if diff := cmp.Diff(expectedModules, loadedModules); diff != "" {
  t.Fatalf("unexpected difference of loaded modules (-want, +got): %v", diff)
}
```

abrender commented on pkg/libinit/root_linux_test.go:
```diff
@@ -0,0 +1,121 @@
+		},
+	}
+
+	dir, err := os.MkdirTemp("", "test_moddir")
```
I think you can replace this whole block with `T.TempDir` (https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff)

abrender commented on pkg/libinit/root_linux_test.go:
```diff
@@ -0,0 +1,121 @@
+	var checkModules []string
+	for _, file := range files {
+		p := filepath.Join(dir, file.Name)
+		os.WriteFile(p, toBytes(file.Content), 0o644)
```
`WriteFile` returns an error, probably a good idea to check it

Foxboron commented on pkg/libinit/root_linux.go:
```diff
@@ -233,25 +235,82 @@ func CreateRootfs() {
-	"idpf":     true,
-	"idpf_imc": true,
+// InitModuleLoader wraps the resources we need for early module loading
+type InitModuleLoader struct {
```
We need some way to specify them, possibly outside of the library itself. So I'd rather keep it exported else we need to have a setter function as well.

Foxboron commented:
I've addressed the feedback from @abrender. Thanks for the review :)

abrender commented on pkg/libinit/root_linux.go:
```diff
@@ -263,24 +320,78 @@ func InstallModules(pattern string, exclude map[string]bool) error {
+			continue
+		} else if strings.HasPrefix(i, "#") {
+			continue
+		} else {
```
I think the `else` and `else if` are not needed because all the previous conditions have `continue`.  So you can either do
```
if i == "" {
  continue
}
if strings.HasPrefix(i, "#") {
  continue
}
modules = append(modules, i)
```

or just combine the `if`s:
```
if i = "" || strings.HasPrefix(i, "#") {
  continue
}
modules = append(modules, i)
```

I prefer the latter but don't have a strong preference.

abrender commented on pkg/libinit/root_linux_test.go:
```diff
@@ -0,0 +1,119 @@
+	for _, file := range files {
+		p := filepath.Join(dir, file.Name)
+		if err := os.WriteFile(p, toBytes(file.Content), 0o644); err != nil {
+			t.Fatal(err)
```
If this t.Fatal triggers there's no indication to the user which of the multiple tests has failed.  You can fix this with "Table-driven tests using subtests" (https://go.dev/blog/subtests):

```
for _, file := range files {
  t.Run(files.Name, func(t *testing.T) {
    // Rest of existing code here
  })
}
```

abrender commented on pkg/libinit/root_linux_test.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	dir := t.TempDir()
+	defer os.RemoveAll(dir)
```
no need to manually remove the dir:  https://pkg.go.dev/testing#B.TempDir says `The directory is automatically removed by Cleanup when the test and all its subtests complete`

Foxboron commented:
Fixed up the last comments.

Should we squash the commits before merging?

abrender commented:
> Fixed up the last comments.

Thanks! I don't see them pushed yet (probably related to your question below)

> 
> Should we squash the commits before merging?

I'll defer to @rminnich.

Foxboron commented:
Sorry :) Amended the commit and everything should be there now

abrender commented:
Awesome, thanks for this contribution! Looks good from me on the code side... I'll leave it to @rminnich for further review/merge.

rminnich commented on pkg/libinit/root_linux.go:
```diff
@@ -233,25 +234,81 @@ func CreateRootfs() {
-var excludedMods = map[string]bool{
-	"idpf":     true,
-	"idpf_imc": true,
+// InitModuleLoader wraps the resources we need for early module loading
```
one little nit: it's frequently best to define an interface, and the create types like this struct that implement the interface. Why? b/v freebsd has modules, just not modules like this. Also, mocking for testing is easier. You can see this in https://github.com/u-root/u-root/blob/main/pkg/efivarfs/vars_test.go#L16, where I was able to put together things that allowed tests to be easy.

rminnich commented:
So I know you need this, and testing the functionality is tricky, since it runs as root and adds modules to the kernel. All that said, you might want to think hard about upping the testing. Approving since it's critical infra.

rminnich commented:
squashing the commits first would be a good idea, I just noticed titles like 'fix up' :-)

rminnich reviewed (changes requested):
makes sense to squash them, then we should be ready to go.

Foxboron commented:
So I understand the need for better test coverage. But it's admittedly a bit hard to do full integration testing since we don't really deal with the `kmod` portion of the code. I'm a tiny bit unsure what more test coverage would get us as we are in practise only really implementing two different parsers.

The commits have been squased :) I'm also around OSFC
